File: tools/merge-static-library.py
#!/usr/bin/env python3
"""Append objects to an existing static library without re-parsing Mach-O."""

from __future__ import annotations

import shutil
import subprocess
import sys
from pathlib import Path


def main() -> None:
    if len(sys.argv) < 4:
        raise SystemExit("usage: merge-static-library.py OUTPUT BASE OBJECT...")

    output = Path(sys.argv[1])
    base = Path(sys.argv[2])
    objects = [Path(argument) for argument in sys.argv[3:]]
    output.parent.mkdir(parents=True, exist_ok=True)
    shutil.copyfile(base, output)
    subprocess.run(["xcrun", "ar", "-q", output, *objects], check=True)
    subprocess.run(["xcrun", "ranlib", output], check=True)


if __name__ == "__main__":
    main()

File: tools/merge-static-library.ts
/** Append object files to an existing static library without parsing Mach-O. */
import { run } from "./lib/home";

const args = process.argv.slice(2);
if (args.length < 3) throw new Error("usage: merge-static-library.ts OUTPUT BASE OBJECT...");
const output = args[0], base = args[1], objects = args.slice(2);
const slash = output.lastIndexOf("/");
if (slash > 0) {
  const made = run(["mkdir", "-p", output.slice(0, slash)]);
  if (made.exitCode !== 0) throw new Error(made.stderr || "cannot create archive output directory");
}
const copied = run(["cp", base, output]);
if (copied.exitCode !== 0) throw new Error(copied.stderr || "cannot copy base archive");
const archived = run(["xcrun", "ar", "-q", output].concat(objects));
if (archived.exitCode !== 0) throw new Error(archived.stderr || "cannot append archive members");
const indexed = run(["xcrun", "ranlib", output]);
if (indexed.exitCode !== 0) throw new Error(indexed.stderr || "cannot index merged archive");
